File: src/resource.cpp
#include "resource.hpp"
#include <cstdlib>
#include <sstream>
#include <stdexcept>

namespace vp {

std::map<std::string, std::shared_ptr<ResourceType>> defaultResourceTypes() {
    std::map<std::string, std::shared_ptr<ResourceType>> types;

    auto tcpport = std::make_shared<ResourceType>();
    tcpport->name = "tcpport";
    tcpport->check = "nc -z localhost ${value}";
    tcpport->counter = true;
    tcpport->start = 3000;
    tcpport->end = 9999;
    types["tcpport"] = tcpport;

    auto vncport = std::make_shared<ResourceType>();
    vncport->name = "vncport";
    vncport->check = "nc -z localhost ${value}";
    vncport->counter = true;
    vncport->start = 5900;
    vncport->end = 5999;
    types["vncport"] = vncport;

    auto serialport = std::make_shared<ResourceType>();
    serialport->name = "serialport";
    serialport->check = "nc -z localhost ${value}";
    serialport->counter = true;
    serialport->start = 9600;
    serialport->end = 9699;
    types["serialport"] = serialport;

    auto dbfile = std::make_shared<ResourceType>();
    dbfile->name = "dbfile";
    dbfile->check = "test -f ${value}";
    dbfile->counter = false;
    dbfile->start = 0;
    dbfile->end = 0;
    types["dbfile"] = dbfile;

    auto socket = std::make_shared<ResourceType>();
    socket->name = "socket";
    socket->check = "test -S ${value}";
    socket->counter = false;
    socket->start = 0;
    socket->end = 0;
    types["socket"] = socket;

    auto datadir = std::make_shared<ResourceType>();
    datadir->name = "datadir";
    datadir->check = "";
    datadir->counter = false;
    datadir->start = 0;
    datadir->end = 0;
    types["datadir"] = datadir;

    auto workdir = std::make_shared<ResourceType>();
    workdir->name = "workdir";
    workdir->check = "";
    workdir->counter = false;
    workdir->start = 0;
    workdir->end = 0;
    types["workdir"] = workdir;

    return types;
}

bool checkResource(const ResourceType& rt, const std::string& value) {
    if (rt.check.empty()) {
        return true; // No check command = always available
    }

    // Interpolate check command
    std::string check = rt.check;
    size_t pos = 0;
    while ((pos = check.find("${value}", pos)) != std::string::npos) {
        check.replace(pos, 8, value);
        pos += value.length();
    }

    // Execute check
    int result = system(check.c_str());

    // Natural command behavior: exit 0 = exists/in-use (not available)
    // exit 1 = free/doesn't exist (available)
    return result != 0; // Resource is available if check command fails
}

std::string allocateResource(std::shared_ptr<State> state, const std::string& rtype, const std::string& requestedValue) {
    auto it = state->types.find(rtype);
    if (it == state->types.end()) {
        throw std::runtime_error("unknown resource type: " + rtype);
    }

    auto rt = it->second;
    std::string value;

    if (rt->counter && requestedValue.empty()) {
        // Auto-increment counter
        int current = state->counters[rtype];
        if (current == 0) {
            current = rt->start;
        }

        bool found = false;
        for (int v = current; v <= rt->end; v++) {
            value = std::to_string(v);
            if (checkResource(*rt, value)) {
                state->counters[rtype] = v + 1;
                found = true;
                break;
            }
        }

        if (!found) {
            std::stringstream ss;
            ss << "no available " << rtype << " in range " << rt->start << "-" << rt->end;
            throw std::runtime_error(ss.str());
        }
    } else {
        // Explicit value requested or non-counter resource
        if (!requestedValue.empty()) {
            value = requestedValue;
        } else {
            throw std::runtime_error("resource type " + rtype + " requires explicit value");
        }

        if (!checkResource(*rt, value)) {
            throw std::runtime_error(rtype + " " + value + " not available");
        }
    }

    return value;
}

} // namespace vp

File: resource.go
package main

import (
	"fmt"
	"os/exec"
	"strconv"
	"strings"
)

// Resource represents an allocated resource
type Resource struct {
	Type  string `json:"type"`  // tcpport|vncport|gpu|license|whatever
	Value string `json:"value"` // "3000" or "/path" or "0"
	Owner string `json:"owner"` // Instance name
}

// ResourceType defines a type of resource with validation
type ResourceType struct {
	Name    string `json:"name"`    // Resource type name
	Check   string `json:"check"`   // Shell command to check availability
	Counter bool   `json:"counter"` // Is this auto-incrementing?
	Start   int    `json:"start"`   // Counter start value
	End     int    `json:"end"`     // Counter end value
}

// DefaultResourceTypes returns the built-in resource types
func DefaultResourceTypes() map[string]*ResourceType {
	return map[string]*ResourceType{
		"tcpport": {
			Name:    "tcpport",
			Check:   "nc -z localhost ${value}", // Exits 0 if in use, 1 if available
			Counter: true,
			Start:   3000,
			End:     9999,
		},
		"vncport": {
			Name:    "vncport",
			Check:   "nc -z localhost ${value}", // Exits 0 if in use, 1 if available
			Counter: true,
			Start:   5900,
			End:     5999,
		},
		"serialport": {
			Name:    "serialport",
			Check:   "nc -z localhost ${value}", // Exits 0 if in use, 1 if available
			Counter: true,
			Start:   9600,
			End:     9699,
		},
		"dbfile": {
			Name:    "dbfile",
			Check:   "test -f ${value}", // Exits 0 if exists, 1 if available
			Counter: false,
		},
		"socket": {
			Name:    "socket",
			Check:   "test -S ${value}", // Exits 0 if exists, 1 if available
			Counter: false,
		},
		"datadir": {
			Name:    "datadir",
			Check:   "", // No check - always available
			Counter: false,
		},
		"workdir": {
			Name:    "workdir",
			Check:   "", // No check - always available (informational resource)
			Counter: false,
		},
	}
}

// AllocateResource allocates a resource of the given type
func AllocateResource(state *State, rtype string, requestedValue string) (string, error) {
	rt := state.Types[rtype]
	if rt == nil {
		return "", fmt.Errorf("unknown resource type: %s", rtype)
	}

	var value string

	if rt.Counter && requestedValue == "" {
		// Auto-increment counter
		current := state.Counters[rtype]
		if current == 0 {
			current = rt.Start
		}

		found := false
		for v := current; v <= rt.End; v++ {
			value = strconv.Itoa(v)
			if CheckResource(rt, value) {
				state.Counters[rtype] = v + 1
				found = true
				break
			}
		}

		if !found {
			return "", fmt.Errorf("no available %s in range %d-%d", rtype, rt.Start, rt.End)
		}
	} else {
		// Explicit value requested or non-counter resource
		if requestedValue != "" {
			value = requestedValue
		} else {
			return "", fmt.Errorf("resource type %s requires explicit value", rtype)
		}

		if !CheckResource(rt, value) {
			return "", fmt.Errorf("%s %s not available", rtype, value)
		}
	}

	return value, nil
}

// CheckResource validates resource availability using the check command
func CheckResource(rt *ResourceType, value string) bool {
	if rt.Check == "" {
		return true // No check command = always available
	}

	// Interpolate check command
	check := strings.ReplaceAll(rt.Check, "${value}", value)

	// Execute check
	cmd := exec.Command("sh", "-c", check)
	err := cmd.Run()
	// Natural command behavior: exit 0 = exists/in-use (not available)
	// exit 1 = free/doesn't exist (available)
	return err != nil // Resource is available if check command fails
}
